### PR TITLE
Rustflags in metadata

### PR DESCRIPTION
**Pull request by Eh2406**
This is a small change that adds Rustflags to the `Metadata`. This means, for example, that if you do a plane build then build with "-C target-cpu=native", you will get 2 copies of the dependencies. Con, more space if you are changing Rustflags. Pro, not rebuilding all the dependencies when you switch back.

Suggested by <https://internals.rust-lang.org/t/idea-cargo-global-binary-cache/9002/31>

### CODE REVIEW

**Comment by dwijnand** [comment]
Nice doc addition too!  The test seems overly complex for what it's testing (does it need dependencies? two of them?).

File: tests/testsuite/freshness.rs [hunk 1]
```diff
@@ -1153,6 +1153,54 @@ fn reuse_shared_build_dep() {
...
+        )
+        .file("a/src/lib.rs", "extern crate b;")
+        .file("b/Cargo.toml", &basic_manifest("b", "0.0.1"))
+        .file("b/src/lib.rs", "")
+        .build();
+
+    p.cargo("build").run();
+    p.cargo("build")
+        .env("RUSTFLAGS", "-C target-cpu=native")
+        .run();
```
Comment (ehuss) [inline tests/testsuite/freshness.rs]: This should probably verify that it is rebuilding.

File: src/cargo/core/compiler/context/compilation_files.rs [hunk 2]
```diff
@@ -12,6 +12,34 @@ use super::{BuildContext, Context, FileFlavor, Kind, Layout, Unit};
...
+/// In general this must include all things that need to be distinguished in different parts of
+/// the same build. This is absolutely required or we override things before
+/// we get chance to use them.
+///
+/// We use a hash because it is an easy way to guarantee
+/// that all the inputs can be converted to a valid path.
+///
+/// This also acts as the main layer of caching provided by Cargo.
+/// For example, we want to cache `cargo build` and `cargo doc` separately, so that running one
+/// does not invalidate the artifacts for the other. We do this by including `profile` in the hash,
```
Comment (ehuss) [inline src/cargo/core/compiler/context/compilation_files.rs]: The word `profile` here isn't quite right for the example.  To distinguish build/check/doc/etc, the `CompileMode` is included (`unit.mode`).  The profile is included to track some settings that differ, for example `panic=abort` changes for dependencies based on whether it is for a test or not.  Dependencies can also be built multiple times and they need to be kept separate for example when build script profile overrides are used.  There are a few other reasons, but I think this is partially captured in the comment in the compute function.

File: src/cargo/core/compiler/context/compilation_files.rs [hunk 3]
```diff
@@ -12,6 +12,34 @@ use super::{BuildContext, Context, FileFlavor, Kind, Layout, Unit};
...
+/// thus the artifacts go in different folders and do not override each other.
+/// If we don't add something that we should have, for this reason, we get the
+/// correct output but rebuild more than is needed.
+///
+/// Some things that need to be tract to ensure the correct output should definitely *not*
```
Comment (ehuss) [inline src/cargo/core/compiler/context/compilation_files.rs]: tract → tracked

**Comment by Eh2406** [comment]
The doc improvements are just a rewrite of how @ehuss explained it to me. I guess it is an example of "the best time to wright documentation is when you are new." I know what I did not understand and where I had looked, he new the anceer.

Good call on the test! (I was under the impression that this would only cache dependencies, hence adding them, I was wrong.)

**Comment by Eh2406** [comment]
@ehuss dose that look better?

**Comment by matklad** [comment]
Cursory comment:

As I understand, there are two bits of information which track "dependencies": the hash of the file and the `Fingerprint`. Historically, `Fingerprint` tracked **everything**, and the hash tacked only subset of the things. I *think* the reasoning for this separation was to enable "automatic gc": if you change some flags, the file names stays the same, and so the old file gets overwritten. 

However, I believe we've broken that "automatic gc" logic: if I today build a crate, and then add `[profiles.dev] opt-level = 1` and build again, I get two sets of deps in `target/debug/deps`. With rust 1.10.0 I get only one set of dependencies. 

So perhaps we can do something more radical here and just *define* the file hash to be exactly the fingerprint?  That should simplify internal representation.

**Comment by Eh2406** [comment]
I think there are still some important things that are not in `metadata`. The example @ehuss gave me (when I asked the same thing) was mtimes of files. If your example change was adding `pub fn nop() {}` to main then the "automatic gc" still works as you describe. It is reasonable to ask if we have gone to far in adding things to `metadata`. However, if we are imagining a global shared target_dir, then the problem of the "automatic gc" cleaning something we need for a different project became much more pressing.

**Comment by alexcrichton** [comment]
I agree that we're not quite ready yet to merge these two hashes, the one for fingerprint calculation wasn't ever designed to be in the filename so it'd at least need a review to make sure it's suitable to always throw all of it in the filename. Incrementally adding more to the filename hash though I think is fine (like this PR)

**Comment by Eh2406** [comment]
So what is this waiting on?

**Comment by michaelwoerister** [comment]
I find this change rather problematic. It keeps breaking things (path remapping, PGO, reproducible builds) because it goes against what rustc's `-Cmetadata` flag is intended for. `-C metadata` flag sets the crate disambiguator in rustc, which is only intended to disambiguate different versions of the same crate. Changing it will change every single symbol name the compiler generates for the crate. The disambiguator identifies that state of the source code, not the output artifacts of the compiler. 

If there is some kind of caching of compiled artifacts, then that layer should take care of parsing `RUSTFLAGS`. Otherwise we'll just keep on adding exceptions.

**Comment by Eh2406** [comment]
I am sorry this is making problems! The only problem where the report has come to my attention is the `--remap-path-prefix`, can you provide links to the discussion of the other problems you mentioned?

> If there is some kind of caching of compiled artifacts

There definitely is, cargo has never rebuilt from scratch at every invocation.

**Comment by michaelwoerister** [comment]
> can you provide links to the discussion of the other problems you mentioned?

I just opened this issue: https://github.com/rust-lang/cargo/issues/7416

> There definitely is, cargo has never rebuilt from scratch at every invocation.

I know. I think that improving caching via different metadata strings has problems though:

- It leads to silent blow-up of the target directory where each small change to `RUSTFLAGS` creates its own crate graph of binaries, with probably quite a few of them "zombie" artifacts because the settings used during experimentation are not re-visited again. 
- It changes the name of output artifacts and symbols, something which other tools and scripts might rely on, especially when debugging something and wanting to inspect things that Cargo and `rustc` are generating.

I'm sorry for complaining about your feature, @Eh2406. You wrote such a nice doc string for it `:)`

**Comment by Eh2406** [comment]
This PR was intended to help unlock work on the "zombie artifacts" problem, unfortunately that work has not yet progressed far enough to clean up the mess.

I do not object to a rewrite or removal if this is making more problems than it is worth.

**Pull request merged**